File: zzDNN-wiki-Legacy-VB-Code/Administration.ascx.vb
'
' DotNetNuke® - http://www.dotnetnuke.com
' Copyright (c) 2002-2013
' by DotNetNuke Corporation
'
' Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated 
' documentation files (the "Software"), to deal in the Software without restriction, including without limitation 
' the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and 
' to permit persons to whom the Software is furnished to do so, subject to the following conditions:
'
' The above copyright notice and this permission notice shall be included in all copies or substantial portions 
' of the Software.
'
' THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED 
' TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL 
' THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF 
' CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER 
' DEALINGS IN THE SOFTWARE.
'


Imports DotNetNuke.Common.Globals
Imports DotNetNuke.Services.Localization
Imports DotNetNuke.Entities.Modules

Namespace DotNetNuke.Modules.Wiki



    Partial Class Administration
        Inherits PortalModuleBase
        Protected WikiSettings As Entities.SettingsInfo
        Protected WikiController As New Entities.SettingsController
#Region " Web Form Designer Generated Code "

        'This call is required by the Web Form Designer.
        <System.Diagnostics.DebuggerStepThrough()> Private Sub InitializeComponent()

        End Sub
        Protected WithEvents ContentEditors As DotNetNuke.UI.UserControls.DualListControl
        Protected WithEvents NotifyRoles As DotNetNuke.UI.UserControls.DualListControl

        'NOTE: The following placeholder declaration is required by the Web Form Designer.
        'Do not delete or move it.
        Private designerPlaceholderDeclaration As System.Object

        Private Sub Page_Init(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Init
            'CODEGEN: This method call is required by the Web Form Designer
            'Do not modify it using the code editor.
            InitializeComponent()
            Framework.jQuery.RequestUIRegistration()
            Framework.jQuery.RequestDnnPluginsRegistration()
        End Sub

#End Region

        Private Sub CtrlPage_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
            'Put user code to initialize the page here
           
            ContentEditors.DataTextField = "Text"
            ContentEditors.DataValueField = "Text"
            NotifyRoles.DataTextField = "Text"
            NotifyRoles.DataValueField = "Text"
            If WikiSettings Is Nothing Then

                WikiSettings = WikiController.GetByModuleID(ModuleId)
                If WikiSettings Is Nothing Then
                    WikiSettings = New Entities.SettingsInfo
                    WikiSettings.ModuleID = -1
                    WikiSettings.ContentEditorRoles = "UseDNNSettings"
                End If
            End If
            If Not IsPostBack Then
                DNNSecurityChk.Checked = WikiSettings.ContentEditorRoles.Equals("UseDNNSettings")
                AllowPageComments.Checked = WikiSettings.AllowDiscussions
                AllowPageRatings.Checked = WikiSettings.AllowRatings
                DefaultCommentsMode.Checked = WikiSettings.DefaultDiscussionMode
                DefaultRatingMode.Checked = WikiSettings.DefaultRatingMode
                NotifyMethodUserComments.Checked = WikiSettings.CommentNotifyUsers

                NotifyMethodCustomRoles.Checked = Not WikiSettings.CommentNotifyRoles.StartsWith("UseDNNSettings;") AndAlso WikiSettings.CommentNotifyRoles.Length > 0
                If NotifyMethodCustomRoles.Checked Then
                    NotifyMethodEditRoles.Checked = WikiSettings.CommentNotifyRoles.Contains(";Edit")
                    NotifyMethodViewRoles.Checked = WikiSettings.CommentNotifyRoles.Contains(";View")
                End If

                BindRights()
                If DNNSecurityChk.Checked = True Then
                    ContentEditors.Visible = False
                    WikiSecurity.Visible = False
                Else
                    ContentEditors.Visible = True
                    WikiSecurity.Visible = True
                End If
                If AllowPageComments.Checked Then
                    Me.ActivateComments.Enabled = True
                    Me.DefaultCommentsMode.Enabled = True
                Else
                    Me.ActivateComments.Enabled = False
                    Me.ActivateComments.Checked = False
                    Me.DefaultCommentsMode.Enabled = False
                    Me.DefaultCommentsMode.Checked = False
                End If

                If AllowPageRatings.Checked Then
                    Me.ActivateRatings.Enabled = True
                    Me.DefaultRatingMode.Enabled = True
                Else
                    Me.ActivateRatings.Enabled = False
                    Me.ActivateRatings.Checked = False
                    Me.DefaultRatingMode.Enabled = False
                    Me.DefaultRatingMode.Checked = False
                End If
                If NotifyMethodCustomRoles.Checked Then
                    NotifyRoles.Visible = True
                    lblNotifyRoles.Visible = True
                    Me.NotifyMethodEditRoles.Enabled = False
                    Me.NotifyMethodViewRoles.Enabled = False
                    Me.NotifyMethodViewRoles.Checked = False
                    Me.NotifyMethodEditRoles.Checked = False
                Else
                    Me.NotifyMethodEditRoles.Enabled = True
                    Me.NotifyMethodViewRoles.Enabled = True

                    NotifyRoles.Visible = False
                    lblNotifyRoles.Visible = False
                End If
            End If
        End Sub


        Private Sub DNNSecurityChk_CheckedChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles DNNSecurityChk.CheckedChanged
            If DNNSecurityChk.Checked = True Then
                ContentEditors.Visible = False
                WikiSecurity.Visible = False
            Else
                ContentEditors.Visible = True
                WikiSecurity.Visible = True
            End If
        End Sub

        Private Sub BindRights()

            ' declare roles
            Dim arrAvailableAuthViewRoles As New ArrayList
            Dim arrAvailableNotifyRoles As New ArrayList

            ' add an entry of All Users for the View roles
            arrAvailableAuthViewRoles.Add(New ListItem("All Users", Common.glbRoleAllUsersName))
            ' add an entry of Unauthenticated Users for the View roles
            arrAvailableAuthViewRoles.Add(New ListItem("Unauthenticated Users", glbRoleUnauthUserName))
            ' add an entry of All Users for the Edit roles

            ' process portal roles
            Dim objRoles As New DotNetNuke.Security.Roles.RoleController
            Dim objRole As DotNetNuke.Security.Roles.RoleInfo
            Dim arrRoles As ArrayList = objRoles.GetPortalRoles(PortalId)
            For Each objRole In arrRoles
                arrAvailableAuthViewRoles.Add(New ListItem(objRole.RoleName, objRole.RoleName))
                arrAvailableNotifyRoles.Add(New ListItem(objRole.RoleName, objRole.RoleName))
            Next

            ' populate view roles

            Dim arrAssignedAuthViewRoles As New ArrayList
            Dim arrAssignedNotifyRoles As New ArrayList

            Dim arrAuthViewRoles As Array
            Dim arrAuthNotifyRoles As Array
            If WikiSettings.ContentEditorRoles = "UseDNNSettings" Then
                arrAuthViewRoles = Split(WikiSettings.ContentEditorRoles, "UseDNNSettings")
            Else
                arrAuthViewRoles = Split(WikiSettings.ContentEditorRoles.Split("|")(0), ";")
            End If
            If WikiSettings.CommentNotifyRoles.StartsWith("UseDNNSettings;") Then
                WikiSettings.CommentNotifyRoles = WikiSettings.CommentNotifyRoles.Replace("UseDNNSettings;", String.Empty)
                arrAuthNotifyRoles = Split(WikiSettings.CommentNotifyRoles, ";")

                For Each s As String In arrAuthNotifyRoles
                    If s = "View" Then
                        NotifyMethodViewRoles.Checked = True
                    ElseIf s = "Edit" Then
                        NotifyMethodEditRoles.Checked = True
                    End If
                Next
            Else
                arrAuthNotifyRoles = Split(WikiSettings.CommentNotifyRoles.Split("|")(0), ";")
            End If

            For Each strRole As String In arrAuthViewRoles
                If strRole <> "" Then
                    For Each objListItem As ListItem In arrAvailableAuthViewRoles
                        If objListItem.Value = strRole Then
                            arrAssignedAuthViewRoles.Add(objListItem)
                            arrAvailableAuthViewRoles.Remove(objListItem)
                            Exit For
                        End If
                    Next
                End If
            Next

            For Each strRole As String In arrAuthNotifyRoles
                If strRole <> "" Then
                    For Each objListItem As ListItem In arrAvailableNotifyRoles
                        If objListItem.Value = strRole Then
                            arrAssignedNotifyRoles.Add(objListItem)
                            arrAvailableNotifyRoles.Remove(objListItem)
                            Exit For
                        End If
                    Next
                End If
            Next

            Dim x As Integer = arrAvailableAuthViewRoles.Count
            Dim y As Integer = arrAssignedAuthViewRoles.Count
            ContentEditors.Available = arrAvailableAuthViewRoles
            ContentEditors.Assigned = arrAssignedAuthViewRoles

            NotifyRoles.Assigned = arrAssignedNotifyRoles
            NotifyRoles.Available = arrAvailableNotifyRoles
        End Sub

        Private Sub SaveButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles SaveButton.Click
            SaveSettings()
            Response.Redirect(NavigateURL(), True)
        End Sub

        Private Sub CancelButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles CancelButton.Click
            Response.Redirect(NavigateURL(), True)
        End Sub

        Private Sub AllowPageComments_CheckedChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles AllowPageComments.CheckedChanged
            If AllowPageComments.Checked Then
                Me.ActivateComments.Enabled = True
                Me.ActivateComments.Checked = True

                Me.DefaultCommentsMode.Enabled = True
            Else
                Me.ActivateComments.Enabled = False
                Me.ActivateComments.Checked = False

                Me.DefaultCommentsMode.Enabled = False
                Me.DefaultCommentsMode.Checked = False
            End If
        End Sub
        Private Sub SaveSettings()
            If DNNSecurityChk.Checked = True Then
                WikiSettings.ContentEditorRoles = "UseDNNSettings"
            Else
                Dim list As String = ";"
                For Each li As ListItem In ContentEditors.Assigned
                    list = list + li.Value + ";"
                Next
                WikiSettings.ContentEditorRoles = list
            End If

            If NotifyMethodCustomRoles.Checked = False Then
                WikiSettings.CommentNotifyRoles = "UseDNNSettings"
                If NotifyMethodEditRoles.Checked = True Then
                    WikiSettings.CommentNotifyRoles = WikiSettings.CommentNotifyRoles & ";Edit"
                End If
                If NotifyMethodViewRoles.Checked = True Then
                    WikiSettings.CommentNotifyRoles = WikiSettings.CommentNotifyRoles & ";View"
                End If
            Else
                Dim list As String = ";"
                For Each li As ListItem In NotifyRoles.Assigned
                    list = list + li.Value + ";"
                Next
                WikiSettings.CommentNotifyRoles = list
            End If

            WikiSettings.AllowDiscussions = AllowPageComments.Checked
            WikiSettings.AllowRatings = AllowPageRatings.Checked
            WikiSettings.DefaultDiscussionMode = DefaultCommentsMode.Checked
            WikiSettings.DefaultRatingMode = DefaultRatingMode.Checked
            WikiSettings.CommentNotifyUsers = NotifyMethodUserComments.Checked

            If WikiSettings.ModuleID = -1 Then
                WikiSettings.ModuleID = ModuleId
                WikiController.Add(WikiSettings)
            Else
                WikiController.Update(WikiSettings)
            End If
            ActivateItems()
        End Sub

        Private Sub ActivateItems()
            If ActivateComments.Checked Or ActivateRatings.Checked Then
                Dim TC As New Entities.TopicController
                Dim alltopics As ArrayList
                alltopics = TC.GetAllByModuleID(Me.ModuleId)
                Dim topic As Entities.TopicInfo
                For Each topic In alltopics
                    If topic.AllowDiscussions = False And ActivateComments.Checked Then
                        topic.AllowDiscussions = True
                    End If
                    If topic.AllowRatings = False And ActivateRatings.Checked Then
                        topic.AllowRatings = True
                    End If
                    TC.Update(topic)
                Next
            End If
        End Sub


        Private Sub AllowPageRatings_CheckedChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles AllowPageRatings.CheckedChanged
            If AllowPageComments.Checked Then
                Me.ActivateRatings.Enabled = True
                Me.ActivateRatings.Checked = True

                Me.DefaultRatingMode.Enabled = True
            Else
                Me.ActivateRatings.Enabled = False
                Me.ActivateRatings.Checked = False

                Me.DefaultRatingMode.Enabled = False
                Me.DefaultRatingMode.Checked = False
            End If
        End Sub

        Private Sub NotifyMethodCustomRoles_CheckedChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles NotifyMethodCustomRoles.CheckedChanged
            If NotifyMethodCustomRoles.Checked Then
                NotifyRoles.Visible = True
                lblNotifyRoles.Visible = True

                Me.NotifyMethodEditRoles.Enabled = False
                Me.NotifyMethodViewRoles.Enabled = False
                Me.NotifyMethodViewRoles.Checked = False
                Me.NotifyMethodEditRoles.Checked = False
            Else
                Me.NotifyMethodEditRoles.Enabled = True
                Me.NotifyMethodViewRoles.Enabled = True
                lblNotifyRoles.Visible = False
                NotifyRoles.Visible = False
            End If
        End Sub
    End Class
End Namespace

File: Views/Administration.ascx.cs
#region Copyright

//--------------------------------------------------------------------------------------------------------
// <copyright file="Administration.ascx.cs" company="DNN Corp®">
//      DNN Corp® - http://www.dnnsoftware.com Copyright (c) 2002-2013 by DNN Corp®
//
//      Permission is hereby granted, free of charge, to any person obtaining a copy of this software and
//      associated documentation files (the "Software"), to deal in the Software without restriction,
//      including without limitation the rights to use, copy, modify, merge, publish, distribute,
//      sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is
//      furnished to do so, subject to the following conditions:
//
//      The above copyright notice and this permission notice shall be included in all copies or
//      substantial portions of the Software.
//
//      THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
//      NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
//      NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
//      DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//      OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
// </copyright>
////------------------------------------------------------------------------------------------------------

#endregion Copyright

using DotNetNuke.Entities.Modules;
using DotNetNuke.Security.Roles;
using DotNetNuke.Services.Exceptions;
using DotNetNuke.Wiki.BusinessObjects;
using DotNetNuke.Wiki.BusinessObjects.Models;
using System;
using System.Collections;
using System.Linq;
using System.Web.UI.WebControls;

namespace DotNetNuke.Wiki.Views
{
    /// <summary>
    /// Administration Partial Class
    /// </summary>
    public partial class Administration : PortalModuleBase
    {
        #region Variables

        private const string StrUseDNNSettings = "UseDNNSettings";

        protected Setting mSettingsModel; ////Data from the WikiSettings Busines Object

        // TODO Do we need this? This is legacy code from VB conversion
        private System.Object designerPlaceholderDeclaration;

        #endregion Variables

        #region Ctor

        /// <summary>
        /// Initializes a new instance of the <see cref="Administration"/> class.
        /// </summary>
        public Administration()
        {
            Load += this.CtrlPage_Load;
            Init += this.Page_Init;
        }

        #endregion Ctor

        #region Events

        /// <summary>
        /// Handles the CheckedChanged event of the AllowPageRatings control.
        /// </summary>
        /// <param name="sender">The source of the event.</param>
        /// <param name="e">The <see cref="System.EventArgs"/> instance containing the event
        /// data.</param>
        protected void AllowPageRatings_CheckedChanged(System.Object sender, System.EventArgs e)
        {
            if (AllowPageComments.Checked)
            {
                this.ActivateRatings.Enabled = true;
                this.ActivateRatings.Checked = true;

                this.DefaultRatingMode.Enabled = true;
            }
            else
            {
                this.ActivateRatings.Enabled = false;
                this.ActivateRatings.Checked = false;

                this.DefaultRatingMode.Enabled = false;
                this.DefaultRatingMode.Checked = false;
            }
        }

        /// <summary>
        /// Handles the CheckedChanged event of the AllowPageComments control.
        /// </summary>
        /// <param name="sender">The source of the event.</param>
        /// <param name="e">The <see cref="System.EventArgs"/> instance containing the event
        /// data.</param>
        protected void AllowPageComments_CheckedChanged(System.Object sender, System.EventArgs e)
        {
            if (AllowPageComments.Checked)
            {
                this.ActivateComments.Enabled = true;
                this.ActivateComments.Checked = true;

                this.DefaultCommentsMode.Enabled = true;
            }
            else
            {
                this.ActivateComments.Enabled = false;
                this.ActivateComments.Checked = false;

                this.DefaultCommentsMode.Enabled = false;
                this.DefaultCommentsMode.Checked = false;
            }
        }

        /// <summary>
        /// Handles the Click event of the CancelButton control.
        /// </summary>
        /// <param name="sender">The source of the event.</param>
        /// <param name="e">The <see cref="EventArgs"/> instance containing the event data.</param>
        protected void CancelButton_Click(object sender, EventArgs e)
        {
            Response.Redirect(DotNetNuke.Common.Globals.NavigateURL(), true);
        }

        /// <summary>
        /// Handles the Load event of the CtrlPage control.
        /// </summary>
        /// <param name="sender">The source of the event.</param>
        /// <param name="e">The <see cref="System.EventArgs" /> instance containing the event
        /// data.</param>
        private void CtrlPage_Load(System.Object sender, System.EventArgs e)
        {
            try
            {
                using (UnitOfWork currentUnitOfWork = new UnitOfWork())
                {
                    var settingsBo = new SettingBO(currentUnitOfWork);

                    ////Put user code to initialize the page here

                    ContentEditors.DataTextField = "Text";
                    ContentEditors.DataValueField = "Text";
                    NotifyRoles.DataTextField = "Text";
                    NotifyRoles.DataValueField = "Text";

                    if (this.mSettingsModel == null)
                    {
                        this.mSettingsModel = settingsBo.GetByModuleID(ModuleId);
                        if (this.mSettingsModel == null)
                        {
                            this.mSettingsModel = new Setting();
                            this.mSettingsModel.ModuleId = -1;
                            this.mSettingsModel.ContentEditorRoles = StrUseDNNSettings;
                        }
                    }

                    if (!IsPostBack)
                    {
                        DNNSecurityChk.Checked = this.mSettingsModel.ContentEditorRoles.Equals(StrUseDNNSettings);
                        AllowPageComments.Checked = this.mSettingsModel.AllowDiscussions;
                        AllowPageRatings.Checked = this.mSettingsModel.AllowRatings;
                        DefaultCommentsMode.Checked = this.mSettingsModel.DefaultDiscussionMode == true;
                        DefaultRatingMode.Checked = this.mSettingsModel.DefaultRatingMode == true;
                        NotifyMethodUserComments.Checked = this.mSettingsModel.CommentNotifyUsers == true;

                        NotifyMethodCustomRoles.Checked =
                            !string.IsNullOrWhiteSpace(this.mSettingsModel.CommentNotifyRoles) &&
                            this.mSettingsModel.CommentNotifyRoles.StartsWith("UseDNNSettings;") && !string.IsNullOrWhiteSpace(this.mSettingsModel.CommentNotifyRoles);
                        if (NotifyMethodCustomRoles.Checked)
                        {
                            NotifyMethodEditRoles.Checked = this.mSettingsModel.CommentNotifyRoles.Contains(";Edit");
                            NotifyMethodViewRoles.Checked = this.mSettingsModel.CommentNotifyRoles.Contains(";View");
                        }

                        // Call the BindRights method
                        this.BindRights();

                        if (DNNSecurityChk.Checked == true)
                        {
                            ContentEditors.Visible = false;
                            WikiSecurity.Visible = false;
                        }
                        else
                        {
                            ContentEditors.Visible = true;
                            WikiSecurity.Visible = true;
                        }

                        if (AllowPageComments.Checked)
                        {
                            this.ActivateComments.Enabled = true;
                            this.DefaultCommentsMode.Enabled = true;
                        }
                        else
                        {
                            this.ActivateComments.Enabled = false;
                            this.ActivateComments.Checked = false;
                            this.DefaultCommentsMode.Enabled = false;
                            this.DefaultCommentsMode.Checked = false;
                        }

                        if (AllowPageRatings.Checked)
                        {
                            this.ActivateRatings.Enabled = true;
                            this.DefaultRatingMode.Enabled = true;
                        }
                        else
                        {
                            this.ActivateRatings.Enabled = false;
                            this.ActivateRatings.Checked = false;
                            this.DefaultRatingMode.Enabled = false;
                            this.DefaultRatingMode.Checked = false;
                        }

                        if (NotifyMethodCustomRoles.Checked)
                        {
                            NotifyRoles.Visible = true;
                            lblNotifyRoles.Visible = true;
                            this.NotifyMethodEditRoles.Enabled = false;
                            this.NotifyMethodViewRoles.Enabled = false;
                            this.NotifyMethodViewRoles.Checked = false;
                            this.NotifyMethodEditRoles.Checked = false;
                        }
                        else
                        {
                            this.NotifyMethodEditRoles.Enabled = true;
                            this.NotifyMethodViewRoles.Enabled = true;

                            NotifyRoles.Visible = false;
                            lblNotifyRoles.Visible = false;
                        }
                    }
                }
            }
            catch (Exception exc)
            {
                Exceptions.ProcessModuleLoadException(this, exc);
            }
        }

        /// <summary>
        /// Handles the CheckedChanged event of the DNN Security Check control.
        /// </summary>
        /// <param name="sender">The source of the event.</param>
        /// <param name="e">The <see cref="System.EventArgs"/> instance containing the event
        /// data.</param>
        protected void DNNSecurityChk_CheckedChanged(System.Object sender, System.EventArgs e)
        {
            if (DNNSecurityChk.Checked == true)
            {
                ContentEditors.Visible = false;
                WikiSecurity.Visible = false;
            }
            else
            {
                ContentEditors.Visible = true;
                WikiSecurity.Visible = true;
            }
        }

        /// <summary>
        /// Handles the CheckedChanged event of the NotifyMethodCustomRoles control.
        /// </summary>
        /// <param name="sender">The source of the event.</param>
        /// <param name="e">The <see cref="System.EventArgs"/> instance containing the event
        /// data.</param>
        protected void NotifyMethodCustomRoles_CheckedChanged(object sender, System.EventArgs e)
        {
            if (NotifyMethodCustomRoles.Checked)
            {
                NotifyRoles.Visible = true;
                lblNotifyRoles.Visible = true;

                this.NotifyMethodEditRoles.Enabled = false;
                this.NotifyMethodViewRoles.Enabled = false;
                this.NotifyMethodViewRoles.Checked = false;
                this.NotifyMethodEditRoles.Checked = false;
            }
            else
            {
                this.NotifyMethodEditRoles.Enabled = true;
                this.NotifyMethodViewRoles.Enabled = true;
                lblNotifyRoles.Visible = false;
                NotifyRoles.Visible = false;
            }
        }

        /// <summary>
        /// Handles the Initialize event of the Page control.
        /// NOTE: The following placeholder declaration is required by the Web Form Designer.
        /// Do not delete or move it.
        /// </summary>
        /// <param name="sender">The source of the event.</param>
        /// <param name="e">The <see cref="System.EventArgs"/> instance containing the event
        /// data.</param>
        private void Page_Init(System.Object sender, System.EventArgs e)
        {
            Framework.jQuery.RequestUIRegistration();
            Framework.jQuery.RequestDnnPluginsRegistration();
        }

        /// <summary>
        /// Handles the Click event of the SaveButton control.
        /// </summary>
        /// <param name="sender">The source of the event.</param>
        /// <param name="e">The <see cref="EventArgs"/> instance containing the event data.</param>
        protected void SaveButton_Click(object sender, EventArgs e)
        {
            this.SaveSettings();
            Response.Redirect(DotNetNuke.Common.Globals.NavigateURL(), true);
        }

        #endregion Events

        #region Methods

        /// <summary>
        /// Activates the items.
        /// </summary>
        /// <param name="currentUnitOfWork">The UnitOfWork.</param>
        private void ActivateItems(UnitOfWork currentUnitOfWork)
        {
            if (ActivateComments.Checked | ActivateRatings.Checked)
            {
                TopicBO topicBo = new TopicBO(currentUnitOfWork);

                var alltopics = topicBo.GetAllByModuleID(this.ModuleId);

                foreach (var topic in alltopics)
                {
                    if (topic.AllowDiscussions == false & ActivateComments.Checked)
                    {
                        topic.AllowDiscussions = true;
                    }

                    if (topic.AllowRatings == false & ActivateRatings.Checked)
                    {
                        topic.AllowRatings = true;
                    }

                    topicBo.Update(topic);
                }
            }
        }

        /// <summary>
        /// Gets the edit rights for user roles and binds them to the respective list control
        /// </summary>
        private void BindRights()
        {
            // declare variables
            ArrayList arrAvailableAuthViewRoles = new ArrayList();
            ArrayList arrAvailableNotifyRoles = new ArrayList();
            ArrayList arrAssignedAuthViewRoles = new ArrayList();
            ArrayList arrAssignedNotifyRoles = new ArrayList();
            Array arrAuthViewRoles = null;
            Array arrAuthNotifyRoles = null;

            // add an entry of All Users for the View roles
            arrAvailableAuthViewRoles.Add(new ListItem("All Users", DotNetNuke.Common.Globals.glbRoleAllUsersName));

            // add an entry of Unauthenticated Users for the View roles
            arrAvailableAuthViewRoles.Add(new ListItem("Unauthenticated Users", DotNetNuke.Common.Globals.glbRoleUnauthUserName));

            // process portal roles
            DotNetNuke.Security.Roles.RoleController objRoles = new DotNetNuke.Security.Roles.RoleController();

            var arrRoles = objRoles.GetPortalRoles(PortalId).OfType<RoleInfo>();
            foreach (var objRole in arrRoles)
            {
                arrAvailableAuthViewRoles.Add(new ListItem(objRole.RoleName, objRole.RoleName));
                arrAvailableNotifyRoles.Add(new ListItem(objRole.RoleName, objRole.RoleName));
            }

            // populate view roles
            if (this.mSettingsModel.ContentEditorRoles.Equals(StrUseDNNSettings))
            {
                arrAuthViewRoles = this.mSettingsModel.ContentEditorRoles.Split(new string[] { StrUseDNNSettings }, StringSplitOptions.RemoveEmptyEntries);
            }
            else
            {
                arrAuthViewRoles = this.mSettingsModel.ContentEditorRoles.Split(
                    new char[] { '|' }, StringSplitOptions.RemoveEmptyEntries)[0]
                    .Split(new char[] { ';' }, StringSplitOptions.RemoveEmptyEntries);
            }

            // populate the notify roles
            if (!string.IsNullOrWhiteSpace(this.mSettingsModel.CommentNotifyRoles))
            {
                if (this.mSettingsModel.CommentNotifyRoles.StartsWith("UseDNNSettings;"))
                {
                    this.mSettingsModel.CommentNotifyRoles = this.mSettingsModel.CommentNotifyRoles.Replace("UseDNNSettings;", string.Empty);
                    arrAuthNotifyRoles = this.mSettingsModel.CommentNotifyRoles.Split(new char[] { ';' }, StringSplitOptions.RemoveEmptyEntries);

                    foreach (string curRole in arrAuthNotifyRoles)
                    {
                        if (curRole.Equals("View"))
                        {
                            NotifyMethodViewRoles.Checked = true;
                        }
                        else if (curRole.Equals("Edit"))
                        {
                            NotifyMethodEditRoles.Checked = true;
                        }
                    }
                }
                else
                {
                    arrAuthNotifyRoles = this.mSettingsModel.CommentNotifyRoles.Split(new char[] { '|' })[0].Split(new char[] { ';' });
                }
            }

            if (arrAuthViewRoles != null)
            {
                foreach (string strRole in arrAuthViewRoles)
                {
                    if (!string.IsNullOrEmpty(strRole))
                    {
                        foreach (ListItem objListItem in arrAvailableAuthViewRoles)
                        {
                            if (objListItem.Value == strRole)
                            {
                                arrAssignedAuthViewRoles.Add(objListItem);
                                arrAvailableAuthViewRoles.Remove(objListItem);
                                break;
                            }
                        }
                    }
                }
            }

            if (arrAuthNotifyRoles != null)
            {
                foreach (string strRole in arrAuthNotifyRoles)
                {
                    if (!string.IsNullOrEmpty(strRole))
                    {
                        foreach (ListItem objListItem in arrAvailableNotifyRoles)
                        {
                            if (objListItem.Value == strRole)
                            {
                                arrAssignedNotifyRoles.Add(objListItem);
                                arrAvailableNotifyRoles.Remove(objListItem);
                                break;
                            }
                        }
                    }
                }
            }

            int x = arrAvailableAuthViewRoles.Count; // TODO Do we need this?
            int y = arrAssignedAuthViewRoles.Count; // TODO Do we need this?
            ContentEditors.Available = arrAvailableAuthViewRoles;
            ContentEditors.Assigned = arrAssignedAuthViewRoles;

            NotifyRoles.Assigned = arrAssignedNotifyRoles;
            NotifyRoles.Available = arrAvailableNotifyRoles;
        }

        /// <summary>
        /// Saves the settings.
        /// </summary>
        private void SaveSettings()
        {
            using (UnitOfWork currentUnitOfWork = new UnitOfWork())
            {
                var settingsBo = new SettingBO(currentUnitOfWork);
                if (DNNSecurityChk.Checked == true)
                {
                    this.mSettingsModel.ContentEditorRoles = StrUseDNNSettings;
                }
                else
                {
                    string list = ";";
                    foreach (ListItem li in ContentEditors.Assigned)
                    {
                        list = list + li.Value + ";";
                    }

                    this.mSettingsModel.ContentEditorRoles = list;
                }

                if (NotifyMethodCustomRoles.Checked == false)
                {
                    this.mSettingsModel.CommentNotifyRoles = StrUseDNNSettings;
                    if (NotifyMethodEditRoles.Checked == true)
                    {
                        this.mSettingsModel.CommentNotifyRoles = this.mSettingsModel.CommentNotifyRoles + ";Edit";
                    }

                    if (this.NotifyMethodViewRoles.Checked == true)
                    {
                        this.mSettingsModel.CommentNotifyRoles = this.mSettingsModel.CommentNotifyRoles + ";View";
                    }
                }
                else
                {
                    string list = ";";
                    foreach (ListItem li in NotifyRoles.Assigned)
                    {
                        list = list + li.Value + ";";
                    }

                    this.mSettingsModel.CommentNotifyRoles = list;
                }

                this.mSettingsModel.AllowDiscussions = AllowPageComments.Checked;
                this.mSettingsModel.AllowRatings = AllowPageRatings.Checked;
                this.mSettingsModel.DefaultDiscussionMode = DefaultCommentsMode.Checked;
                this.mSettingsModel.DefaultRatingMode = DefaultRatingMode.Checked;
                this.mSettingsModel.CommentNotifyUsers = NotifyMethodUserComments.Checked;

                if (this.mSettingsModel.ModuleId == -1)
                {
                    this.mSettingsModel.ModuleId = ModuleId;
                    settingsBo.Add(this.mSettingsModel);
                }
                else
                {
                    settingsBo.Update(this.mSettingsModel);
                }

                this.ActivateItems(currentUnitOfWork);
            }
        }

        /// <summary>
        /// Binds role controls.
        /// </summary>
        /// <param name="roleType">Type of the role.</param>
        private void zzBindRoleControls(string roleType)
        {
            // declare variables
            string roles = string.Empty;
            ListItem[] arrAuthRoles = null;
            ArrayList arrAssignedRoles = new ArrayList();
            ArrayList arrAvailableRoles = new ArrayList();

            if (roleType == "ContentEditors")
            {
                roles = this.mSettingsModel.ContentEditorRoles;
            }
            else if (roleType == "CommentNotifyRoles")
            {
                roles = this.mSettingsModel.CommentNotifyRoles;
            }
            else
            {
                // TODO We've got problems if its not "ContentEditors" or "CommentNotifyRoles" TODO
                // Need to raise an error
            }

            if (roles != StrUseDNNSettings)
            {
                arrAuthRoles = roles.Split(
                    new char[] { '|' }, StringSplitOptions.RemoveEmptyEntries)[0]
                    .Split(new char[] { ';' }, StringSplitOptions.RemoveEmptyEntries).Select(p =>
                    new ListItem(p, p)).ToArray();

                // Convert the arrAuthRoles array to an array list
                arrAssignedRoles.AddRange(arrAuthRoles);
            }

            if (roleType == "CommentNotifyRoles")
            {
                foreach (ListItem curRole in arrAssignedRoles)
                {
                    if (curRole.Value == string.Empty)
                    {
                        arrAvailableRoles.Remove(curRole);
                    }
                }
            }

            // Call the BuildAllRolesArray method to Build an Array of All Roles ----------------
            // Available Roles = All Roles - Assigned Roles AvailableRoles(arrAssignedRoles);

            // Build Available Roles add an entry of All Users for the View roles
            arrAvailableRoles.Add(new ListItem("All Users", DotNetNuke.Common.Globals.glbRoleAllUsersName));

            // add an entry of Unauthenticated Users for the View roles
            arrAvailableRoles.Add(new ListItem("Unauthenticated Users", DotNetNuke.Common.Globals.glbRoleUnauthUserName));

            // process portal roles
            DotNetNuke.Security.Roles.RoleController objRoles = new DotNetNuke.Security.Roles.RoleController();

            var arrRoles = objRoles.GetPortalRoles(PortalId).OfType<RoleInfo>();
            foreach (var objRole in arrRoles)
            {
                arrAvailableRoles.Add(new ListItem(objRole.RoleName, objRole.RoleName));
            }

            // Remove the Assigned Roles from the Available Roles
            if (arrAvailableRoles.Count > 0)
            {
                foreach (ListItem curRole in arrAssignedRoles)
                {
                    arrAvailableRoles.Remove(curRole);
                }
            }

            if (roleType == "ContentEditors")
            {
                ContentEditors.Assigned = arrAssignedRoles;
                ContentEditors.Available = arrAvailableRoles;
            }
            else if (roleType == "CommentNotifyRoles")
            {
                NotifyRoles.Assigned = arrAssignedRoles;
                NotifyRoles.Available = arrAvailableRoles;
            }
            else
            {
                // TODO We've got problems if its not "ContentEditors" or "CommentNotifyRoles" TODO
                // Need to raise an error
            }
        }

        /// <summary>
        /// Get the available roles.
        /// </summary>
        /// <param name="arrAssignedRoles">An array of assigned roles.</param>
        /// <returns>ArrayList of Available Roles</returns>
        private ArrayList zzAvailableRoles(ArrayList arrAssignedRoles)
        {
            // Declare Variables
            ArrayList arrAvailableRoles = new ArrayList();

            // add an entry of All Users for the View roles
            arrAvailableRoles.Add(new ListItem("All Users", DotNetNuke.Common.Globals.glbRoleAllUsersName));

            // add an entry of Unauthenticated Users for the View roles
            arrAvailableRoles.Add(new ListItem("Unauthenticated Users", DotNetNuke.Common.Globals.glbRoleUnauthUserName));

            // process portal roles
            DotNetNuke.Security.Roles.RoleController objRoles = new DotNetNuke.Security.Roles.RoleController();

            var arrRoles = objRoles.GetPortalRoles(PortalId).OfType<RoleInfo>();
            foreach (var objRole in arrRoles)
            {
                arrAvailableRoles.Add(new ListItem(objRole.RoleName, objRole.RoleName));
            }

            // Remove the Assigned Roles from the Available Roles
            if (arrAvailableRoles.Count > 0)
            {
                foreach (ListItem curRole in arrAssignedRoles)
                {
                    arrAvailableRoles.Remove(curRole);
                }
            }

            return arrAvailableRoles;
        }

        #endregion Methods
    }
}